### PR TITLE
KAFKA-6145: KIP-441 Pt. 6 Trigger probing rebalances until group is stable

### PR DESCRIPTION
**Pull request by ableegoldman**
This KIP is fairly straightforward, and does as the title describes by enforcing a rebalance once the configured `probing.rebalance.interval` has elapsed. However, we have had to modify the original plan in the KIP slightly to handle an edge case with static membership enabled:

Since the group leader can crash and restart without triggering a rebalance, we can't rely on a purely in-memory flag/counter to keep track of these probing rebalances. We can instead rely on the assignment, encoding the upcoming probing rebalance in the `AssignmentInfo`. This is encoded as the time in ms of the next scheduled rebalance, ie it is set to `currentTimeMs + probingRebalanceIntervalMs` when the assignment is being generated. This anchors the probing rebalances to wall clock time, and ensures a pathologically failing member will not prevent the group from ever rebalancing. We leave it up to a single member to be responsible for triggering the probing rebalances, and encode this for a single consumer on the leader's client (chosen arbitrarily). 

This PR also adds tracking of whether a followup rebalance is required to `StreamsPartitionAssignor#assign` in order to log when the group is actually stable. This applies not only to KIP-441, but to version probing and cooperative rebalances as well.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Q: This does'nt seem to be used anywhere. Should it be removed?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: prop:
```suggestion
        final boolean followupRebalanceNeeded = 
            assignTasksToClients(allSourceTopics, partitionsForTask, topicGroups, clientMetadataMap, fullMetadata);
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: req: This comments is not correct anymore, isn't it? Please correct it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:367]: req: Those steps comment do not match the steps in the description of the method. Please make them match.

**Review by cadonna — COMMENTED** [review]
@ableegoldman Thank you for the PR!

Here my first comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Nope, but it should have been `AtomicLong`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]:
```suggestion
                } else if (nextProbingRebalanceMs.get() < Time.milliseconds()) {
```

This will let us control it in testing, which we should do.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]:
```suggestion
                                "rebalance to probe for newly caught-up clients");
```

"aught" is so 2000s.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:367]: done

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Ideally, we'd use the `Time#milliseconds()` here instead of `System.currentTimeMillis()`, but you'd have to pass the `time` instance in through `configure`. It would aid in testability, but it's not strictly necessary. I'll leave it to you.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1023]: Por favor, `else { "Finished unstable assignment of tasks. Followup rebalance is scheduled for {time}." }`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java:361]: I'm wondering if we can avoid encoding this for the common case where there is no "next rebalance time". Do you have thoughts on the subject?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java]: Am I right in thinking that all that happened to these older ones is that you added line breaks?

Can we do that in a later PR that doesn't also change the production logic, so it's easier to see that we're not invalidating the tests here? I.e., it's suspicious to change regression tests while you're changing the logic the regression tests would be checking.

**Review by vvcephei — COMMENTED** [review]
Thanks for this! It looks good overall, but I had a few comments.

**Comment by vvcephei** [comment]
test this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Seems worth it

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1023]: It doesn't make sense to add the time here, as the assignment can be "unstable" for other reasons (partitions to be revoked, version probing). But we can log the time if it is a probing rebalance, and just say "unstable assignment" here

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java:361]: I think it's only worth doing if we don't intend to switch over to generated protocol code for the `AssignmnentInfo` anytime soon. I'm pretty sure it's not possible to accomplish this in the json schema, so it would be a short lived optimization and/or require us to bump the protocol version unnecessarily once we do switch over.

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java:361]: Of course, we don't _have_ to move the `AssignmentInfo` to generated code as well (although imo it's a bit weird to do it for the subscription but not the assignment). Or we could probably just add support for this ourselves at some later point.

Do you think we should try and save these 8 bytes per assignment?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java]: Yeah it's just fixing the spacing, I'll remove it

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1023]: Sounds good. The time thing was just me playing jazz. The main thing was that it seems asymmetrical to only log when we do a stable assignment; it seems nice as an operator to also see that we finished an unstable assignment and that we'll stabilize it later.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java:361]: Yeah, this is a good reason to hold off. We can always bump the version later to try a different encoding.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java]: Thanks!

**Comment by vvcephei** [comment]
test this please

**Comment by ableegoldman** [comment]
System tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3880/

**Comment by vvcephei** [comment]
Thanks, @ableegoldman ! There were still a few unfortunately reformatted lines in that AssignmentInfoTest, but it's close enough now for me to feel confident about it. I might fix it while merging. 

Thanks for running the system tests; I think this is ready to merge once the integration tests and system tests pass.

**Comment by ableegoldman** [comment]
Java8 build passed, Java11 failed with unrelated flaky `kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup`

**Comment by ableegoldman** [comment]
System tests all passed: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2020-04-06--001.1586227165--ableegoldman--KIP-441-trigger-rebalances-until-stable--43f2304/report.html

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: prop:
```suggestion
        consumerConfigs.put(StreamsConfig.InternalConfig.TIME, time);
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]:
```suggestion
     *    they have been created with the right number of partitions. Also verify and/or create
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:693]: prop: Please fix indentation of parameters.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]:
```suggestion
            if (populateActiveTaskAndPartitionsLists(
                    activePartitionsList,
                    assignedActiveList,
                    consumer,
                    clientMetadata.state,
                    activeTasksForConsumer,
                    partitionsForTask,
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]:
```suggestion
                    allOwnedPartitions
                )) {
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java:192]: req: Could you please add unit tests for these two methods? I am aware that there isn't even a `AssignorConfigurationTest` class (which makes me a bit sad ;-)). We can add the unit tests for the other methods in a separate PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:369]: req: Please add a test that calls `StreamThread#create()` with a custom `KafkaClientSupplier` that provides a mock consumer and verify that the internal config is set.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:566]: req: I guess you could also verify this with a mock consumer.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java]: super-nit: `nextRebalanceTimeShouldBeMxValueByDefault()` -> `nextRebalanceTimeShouldBeMaxValueByDefault()`

**Review by cadonna — COMMENTED** [review]
@ableegoldman Thank you for the PR!

Here my feedback

**Comment by vvcephei** [comment]
test this please

**Comment by vvcephei** [comment]
test this please

**Comment by vvcephei** [comment]
test this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:369]: This feels a bit out of scope, since we don't currently have _any_ tests that verify `StreamThread#create()`. Also, failing to set this attribute would instantly break all of our integration tests, since the consumer wouldn't be able to instantiate the assignor.

Not saying it wouldn't be valuable to have a unit test for it, but maybe we can call this one a follow-on task.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:566]: added `shouldEnforceRebalanceAfterNextScheduledProbingRebalanceTime`

**Comment by vvcephei** [comment]
test this please

**Comment by vvcephei** [comment]
test this please

**Comment by vvcephei** [comment]
Here's the failure:
```
00:05:11.393 > Task :streams:compileTestJava
00:05:11.393 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:484: error: cannot find symbol
00:05:11.393         mockClientSupplier.setClusterForAdminClient(createCluster());
00:05:11.393                           ^
00:05:11.394   symbol:   method setClusterForAdminClient(Cluster)
00:05:11.394   location: variable mockClientSupplier of type EasyMockConsumerClientSupplier
```

**Comment by ableegoldman** [comment]
@vvcephei Am I going crazy, or did Jenkins pick up the wrong commit or something? Note that the line numbers in the failure don't match up with the line numbers in `StreamThreadTest` under `Files Changed`. FWIW I've run it several times locally as well

**Comment by vvcephei** [comment]
Hey @ableegoldman , Jenkins merges with trunk before running the tests, so it probably picked up a change that's happened in trunk since you branched.

**Comment by ableegoldman** [comment]
Oh, yep, duh. Guess that qualifies as stupid, not crazy 😄 Rebased on trunk and pushed the fix, can you kick off tests again?

**Comment by vvcephei** [comment]
test this please

**Comment by vvcephei** [comment]
test this please

**Comment by vvcephei** [comment]
retest this please

**Comment by vvcephei** [comment]
Haha, neither stupid nor crazy. I remember very clearly when I learned that our Jenkins build does this, because I also felt like I was going nuts.

**Comment by ableegoldman** [comment]
One unrelated failure: `kafka.api.PlaintextAdminIntegrationTest.testAlterReplicaLogDirs`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:369]: Agreed

**Comment by cadonna** [comment]
> Haha, neither stupid nor crazy. I remember very clearly when I learned that our Jenkins build does this, because I also felt like I was going nuts.

Same here

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1869]: I was more thinking about a separate unit test class `AssignorConfigurationTest` with a more thorough testing of all code paths, but if we feel that this should also be a follow up task, that is also fine with me.

**Review by cadonna — APPROVED** [review]
LGTM

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1869]: Thanks for pointing these out, @cadonna .

**Comment by vvcephei** [comment]
Looks like we're good to go. Thanks, all!

**Pull request merged**
